Sources_v3/Workers/MessageUpdater.swift

```diff
@@ -0,0 +1,109 @@
+//
+// Copyright © 2020 Stream.io Inc. All rights reserved.
+//
+
+import CoreData
+import Foundation
+
+class MessageUpdater<ExtraData: ExtraDataTypes>: Worker {
+    func getMessage(cid: ChannelId, messageId: MessageId, completion: ((Error?) -> Void)? = nil) {
+        let endpoint: Endpoint<MessagePayload<ExtraData>> = .getMessage(messageId: messageId)
+        apiClient.request(endpoint: endpoint) {
+            switch $0 {
+            case let .success(message):
+                self.database.write({ session in
+                    try session.saveMessage(payload: message, for: cid)
+                }, completion: { error in
+                    completion?(error)
+                })
+            case let .failure(error):
+                completion?(error)
+            }
+        }
+    }
+    
+    func deleteMessage(messageId: MessageId, completion: ((Error?) -> Void)? = nil) {
+        var shouldDeleteOnBackend = true
+        
+        database.write({ session in
+            guard
+                let dto = session.message(id: messageId), [.pendingSend, .sendingFailed].contains(dto.localMessageState)
+            else { return }
+            
+            session.delete(dto)
+            shouldDeleteOnBackend = false
+        }, completion: { error in
+            guard shouldDeleteOnBackend, error == nil else {
+                completion?(error)
+                return
+            }
+            
+            self.database.write({ session in
+                session.message(id: messageId)?.localMessageState = .deleting
```

Comment:
I think you can do this already in the session above, you know the state will change to `deleting` if the message exists on the backend.